core/sessions/ldapauth/reaper.go

```diff
@@ -0,0 +1,160 @@
+	err := ldSync.deleteStaleSessions(recordCreationStaleThreshold)
+	if err != nil {
+		ldSync.lggr.Error("unable to reap stale sessions: ", err)
+	}
+
+	// For each defined role/group, query for the list of group members to gather the full list of possible users
+	users := []sessions.User{}
+
+	// Establish ephemeral connection
+	conn, err := ldap.DialURL(ldSync.config.ServerAddress())
+	if err != nil {
+		ldSync.lggr.Errorf("Failed to Dial LDAP Server", err)
+	}
+	// Root level root user auth with credentials provided from config
+	bindStr := ldSync.config.BaseUserAttr() + "=" + ldSync.config.ReadOnlyUserLogin() + "," + ldSync.config.BaseDN()
+	if err := conn.Bind(bindStr, ldSync.config.ReadOnlyUserPass()); err != nil {
+		ldSync.lggr.Errorf("Unable to login as initial root LDAP user", err)
+	}
+	defer conn.Close()
+
+	// Query for list of uniqueMember IDs present in Admin group
+	adminUsers, err := ldSync.ldapGroupMembersListToUser(conn, ldSync.config.AdminUserGroupCN(), sessions.UserRoleAdmin)
+	if err != nil {
+		ldSync.lggr.Errorf("Error in ldapGroupMembersListToUser: ", err)
+	}
+	// Query for list of uniqueMember IDs present in Edit group
+	editUsers, err := ldSync.ldapGroupMembersListToUser(conn, ldSync.config.EditUserGroupCN(), sessions.UserRoleEdit)
+	if err != nil {
+		ldSync.lggr.Errorf("Error in ldapGroupMembersListToUser: ", err)
+	}
+	// Query for list of uniqueMember IDs present in Edit group
+	runUsers, err := ldSync.ldapGroupMembersListToUser(conn, ldSync.config.RunUserGroupCN(), sessions.UserRoleRun)
+	if err != nil {
+		ldSync.lggr.Errorf("Error in ldapGroupMembersListToUser: ", err)
+	}
+	// Query for list of uniqueMember IDs present in Edit group
+	readUsers, err := ldSync.ldapGroupMembersListToUser(conn, ldSync.config.ReadUserGroupCN(), sessions.UserRoleView)
+	if err != nil {
+		ldSync.lggr.Errorf("Error in ldapGroupMembersListToUser: ", err)
+	}
+
+	users = append(users, adminUsers...)
+	users = append(users, editUsers...)
+	users = append(users, runUsers...)
+	users = append(users, readUsers...)
+
+	// Dedupe preserving order of highest role
+	uniqueRef := make(map[string]struct{})
+	upstreamUserState := []sessions.User{}
+	for _, user := range users {
+		if _, ok := uniqueRef[user.Email]; !ok {
+			uniqueRef[user.Email] = struct{}{}
+			upstreamUserState = append(upstreamUserState, user)
+		}
+	}
+
+	// upstreamUserState is now the most up to date source of truth
+	// Update state of local ldap tables, purging users not present in
+	// up to date list, and downgrading roles where applicable
+	err = ldSync.q.Transaction(func(tx pg.Queryer) error {
+		emailList := []interface{}{}
+		for _, user := range upstreamUserState {
+			emailList = append(emailList, user.Email)
+		}
+		placeholders := make([]string, len(emailList))
+		for i := range emailList {
+			placeholders[i] = "?"
+		}
+		query := fmt.Sprintf("DELETE FROM ldap_sessions WHERE email NOT IN (%s)", strings.Join(placeholders, ", "))
```

Comment:
Is it possible for len(placeholders) to be very large? I don't think this style of query scales very well in that case :thinking:

Reply:
Oh interesting, is that a problem when using the query library this way? What is the alternative if there are a lot of parameters? For LDAP auth it will be advised that the session duration be set low. With a couple concurrent users this table shouldn't have more than a dozen at a one time

Reply:
@jmank88 Ah I see - that's a really good point. I've pushed a new commit to invert the logic here, since the implementation does pull the full list/state of upstream users so that we can sync the local database. 

Updated now!